geode-core/src/main/java/org/apache/geode/cache/configuration/RegionConfig.java

```diff
@@ -678,13 +685,7 @@ public void setDeclarable(DeclarableType value) {
    *
    */
   @XmlAccessorType(XmlAccessType.FIELD)
-  @XmlType(name = "", propOrder = {"functional", "primaryKey"})
-  public static class Index {
-
-    @XmlElement(namespace = "http://geode.apache.org/schema/cache")
-    protected RegionConfig.Index.Functional functional;
-    @XmlElement(name = "primary-key", namespace = "http://geode.apache.org/schema/cache")
-    protected RegionConfig.Index.PrimaryKey primaryKey;
+  public static class Index implements CacheElement {
```

Comment:
As I mentioned in your other PR, I don't see anywhere indicating that `functional` or `primary-key` are deprecated.  It seems hasty to remove them now.

Reply:
Does this implement `CacheElement` only so it can be set in the `setCacheElement` and consumed in the `updateClusterConfig`?

The conversation on the dev list didn't get any traction, but I think the scope of `CacheElement` should be kept to third-party extension configuration classes.  If the `updateClusterConfig` call is going to require a cast every time anyway, there's no harm in having that signature accept an `Object`.  For finding internal configuration objects in the `clusterConfig`, it would be easy to pass the identifying method along with the object, a la
```
   static <T extends CacheElement> T findElement(List<T> list, String id) {
    return findElement(list, id, T::getId);
  }

  static <T> T findElement(List<T> list, String targetId, Function<T, String> identifyBy) {
    return list.stream().filter(o -> identifyBy.apply(o).equals(targetId)).findFirst().orElse(null);
   }
```
and, if we ever need it, using `findElement(cacheConfig.getIndices(), indexName, Index::getName);`

Just a thought.

Reply:
The Index implement CacheElement because 
1) we will need to be able search by it's name, so it needs to be Identifiable
2) it's used in a function argument, so it needs to be Serializable.

If you view CacheElement just a combination of these two interfaces (something identifiable and serializable), anything that needs to be both can have this interface.  a 3rd party custom element will fall into this category and maybe more, but it should be at least this.

Reply:
With respect to the first bit, I'm just saying that we should identify if this is a dead corner of the code and, if it is, properly deprecate it for removal.  But yeah, maybe that's something for the query guys.

With respect to the second half, I think just using `CacheElement` as a convenient combination of two interfaces is a bit of a trap.  If the label itself doesn't have meaning, then I don't think it belongs -- we should just use `Identifiable<String>` and `Serializable` themselves.  If it were just a combination of two interfaces, then a class like `BootstrappingFunction` is technically a `CacheElement`, since it implements both of these interfaces.

In this context, I could see either side of `Index` getting the label.  I would just really like a proper definition of what, conceptually, the `CacheElement` is meant to signify, beyond what it is at an implementation level.

Reply:
How about this definition:
Any element that's in CacheConfig that needs to be identifiable and serializable?

Reply:
That sounds grand.

How do we determine which need to be Identifiable?  Anything with a `name` or `id` field?  Anything at the top-level?  Anything that ever needs to be modified, implying every class within the configuration?

Reply:
when we need to be able to search by the name/id in the command. again, it doesn't have to be an all-or-nothing thing. when we refactor the commands, we will see which ones are necessary.